src/Microsoft.VisualStudio.Threading.Shared/AsyncBarrier.cs

```diff
@@ -52,21 +50,33 @@ public AsyncBarrier(int participants)
         /// <returns>An awaitable.</returns>
         public Task SignalAndWait()
         {
-            var tcs = new TaskCompletionSource<bool>();
-            this.waiters.Push(tcs);
-            if (Interlocked.Decrement(ref this.remainingParticipants) == 0)
+            lock (this.waiters)
```

Comment:
I hear you on that. We know `AsyncSemaphore` use in Roslyn is suffering from contention within `SemaphoreSlim`, and I have sent Heejae and Cyrus a build from a branch where I fix that (or at least, I hope I fix it) to see how effective it is. 
I'm just not prioritizing this one because we literally have 0 uses of it across every repo I have access to. And I'm also not convinced that a briefly held lock is always worse than an Interlocked.CompareExchange with retry loop, yet the latter is more complicated and warrants more thorough testing. So I'd rather be data driven before introducing such a pattern here.